integration/src/integrationtests/java/com/aws/greengrass/integrationtests/ClientDeviceAuthorizerIntegrationTest.java

```diff
@@ -0,0 +1,142 @@
+import java.nio.file.Path;
+import java.security.KeyStoreException;
+import java.security.UnrecoverableKeyException;
+import java.security.cert.X509Certificate;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+import static com.aws.greengrass.testcommons.testutilities.ExceptionLogProtector.ignoreExceptionOfType;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.mockito.Mockito.when;
+
+@ExtendWith({GGExtension.class, MockitoExtension.class})
+public class ClientDeviceAuthorizerIntegrationTest {
+    private Kernel kernel;
+    private static final String DEFAULT_CLIENT = "myThing";
+    private static final String DEFAULT_TOPIC = "topic";
+    private static final byte[] DEFAULT_PASSWORD = "".getBytes(StandardCharsets.UTF_8);
+    private static final long TEST_TIME_OUT_SEC = 30L;
+    @TempDir
+    Path rootDir;
+    private final Optional<String> certId = Optional.of("certId");
+    private String clientPem;
+    @Mock
+    IotAuthClient iotAuthClient;
+
+    @BeforeEach
+    void setup(ExtensionContext context) throws Exception {
+        ignoreExceptionOfType(context, SpoolerStoreException.class);
+        ignoreExceptionOfType(context, NoSuchFileException.class); // Loading CA keystore
+        ignoreExceptionOfType(context, NullPointerException.class); // Uploading Core Device CA
+        ignoreExceptionOfType(context, KeyStoreException.class);
+        ignoreExceptionOfType(context, UnrecoverableKeyException.class);
+
+        // Set this property for kernel to scan its own classpath to find plugins
+        System.setProperty("aws.greengrass.scanSelfClasspath", "true");
+        startNucleusWithConfig();
+
+        List<X509Certificate> clientCertificates = CertificateTestHelpersMoquette.createClientCertificates(1);
+        this.clientPem = CertificateHelper.toPem(clientCertificates.get(0));
+    }
+
+    @AfterEach
+    void cleanup() {
+        kernel.shutdown();
+    }
+
+    private void startNucleusWithConfig() throws InterruptedException {
+        kernel = new Kernel();
+        kernel.getContext().put(IotAuthClient.class, iotAuthClient);
+
+        CountDownLatch serviceRunning = new CountDownLatch(1);
+        kernel.parseArgs("-r", rootDir.toAbsolutePath().toString(), "-i",
+            getClass().getResource("config.yaml").toString());
+        GlobalStateChangeListener listener = (GreengrassService service, State was, State newState) -> {
+            if (service.getName().equals(MQTTService.SERVICE_NAME) && service.getState().equals(State.RUNNING)) {
+                serviceRunning.countDown();
+            }
+        };
+        kernel.getContext().addGlobalStateChangeListener(listener);
+        kernel.launch();
+        assertTrue(serviceRunning.await(TEST_TIME_OUT_SEC, TimeUnit.SECONDS));
+        kernel.getContext().removeGlobalStateChangeListener(listener);
+    }
+
+    @Test
```

Comment:
Which of these tests _fail_ without your code change?

Reply:
The first test fails without out the code change. The second test succeeds without the code change, but does so by creating a session in CDA and not in Moquette. This means each subsequent `canRead` or `canWrite` will create a new session in CDA without cleaning up the old session. The code change makes it so the second test will create a session in both CDA and Moquette, meaning that we will not be forgetting to clean up sessions in CDA. I can rework the second test so that it fails without the code change.

Reply:
That's ok. 

Why does the first test fail? It is calling `checkValid` which should find the existing session right? Why would it close the existing session?

Reply:
Nevermind, I see.